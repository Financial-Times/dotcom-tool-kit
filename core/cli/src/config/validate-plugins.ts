import { RawConfig, ValidPluginsConfig } from '@dotcom-tool-kit/types'
import { Validated, reduceValidated } from '@dotcom-tool-kit/validated'

export function validatePlugins(config: RawConfig): Validated<ValidPluginsConfig> {
  const validatedPlugins = reduceValidated(
    Object.entries(config.plugins).map(([id, plugin]) => plugin.map((p) => [id, p] as const))
  )
  return validatedPlugins.map((plugins) => ({ ...config, plugins: Object.fromEntries(plugins) }))
}
