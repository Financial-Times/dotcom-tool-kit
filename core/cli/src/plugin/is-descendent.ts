import { Plugin } from '@dotcom-tool-kit/types'

export function isDescendent(possibleAncestor: Plugin, possibleDescendent: Plugin): boolean {
  if (!possibleDescendent.parent) {
    return false
  } else if (possibleDescendent.parent === possibleAncestor) {
    return true
  } else {
    return isDescendent(possibleAncestor, possibleDescendent.parent)
  }
}
