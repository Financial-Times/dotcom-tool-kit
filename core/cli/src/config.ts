import path from 'path'
import type { Logger } from 'winston'

import type { HookTask } from './hook'
import { loadPlugin, resolvePlugin } from './plugin'
import { Conflict, findConflicts, withoutConflicts, isConflict } from './conflict'
import { ToolKitConflictError, ToolKitError } from '@dotcom-tool-kit/error'
import { TaskClass, Hook, mapValidated, Plugin, reduceValidated, Validated } from '@dotcom-tool-kit/types'
import {
  formatTaskConflicts,
  formatUndefinedHookTasks,
  formatUnusedOptions,
  formatHookTaskConflicts,
  formatHookConflicts,
  formatOptionConflicts,
  formatUninstalledHooks,
  formatMissingTasks
} from './messages'

export interface PluginOptions {
  options: Record<string, unknown>
  plugin: Plugin
  forPlugin: Plugin
}

export interface RawConfig {
  root: string
  plugins: { [id: string]: Validated<Plugin> }
  resolvedPlugins: Set<Plugin>
  tasks: { [id: string]: TaskClass | Conflict<TaskClass> }
  hookTasks: { [id: string]: HookTask | Conflict<HookTask> }
  options: { [id: string]: PluginOptions | Conflict<PluginOptions> | undefined }
  hooks: { [id: string]: Hook<unknown> | Conflict<Hook<unknown>> }
}

export type ValidPluginsConfig = Omit<RawConfig, 'plugins'> & {
  plugins: { [id: string]: Plugin }
}

export interface ValidConfig extends ValidPluginsConfig {
  tasks: { [id: string]: TaskClass }
  hookTasks: { [id: string]: HookTask }
  options: { [id: string]: PluginOptions }
  hooks: { [id: string]: Hook<unknown> }
}

const coreRoot = path.resolve(__dirname, '../')

export const createConfig = (): RawConfig => ({
  root: coreRoot,
  plugins: {},
  resolvedPlugins: new Set(),
  tasks: {},
  hookTasks: {},
  options: {},
  hooks: {}
})

async function asyncFilter<T>(items: T[], predicate: (item: T) => Promise<boolean>): Promise<T[]> {
  const results = await Promise.all(items.map(async (item) => ({ item, keep: await predicate(item) })))

  return results.filter(({ keep }) => keep).map(({ item }) => item)
}

export function validateConfig(config: ValidPluginsConfig): asserts config is ValidConfig {
  const hookTaskConflicts = findConflicts(Object.values(config.hookTasks))
  const hookConflicts = findConflicts(Object.values(config.hooks))
  const taskConflicts = findConflicts(Object.values(config.tasks))
  const optionConflicts = findConflicts(Object.values(config.options))

  const definedHookTaskConflicts = hookTaskConflicts.filter((conflict) => {
    return conflict.conflicting[0].id in config.hooks
  })

  let shouldThrow = false
  const error = new ToolKitConflictError(
    'There are problems with your Tool Kit configuration.',
    hookTaskConflicts.map((conflict) => ({
      hook: conflict.conflicting[0].id,
      conflictingTasks: conflict.conflicting.flatMap((hook) =>
        hook.tasks.map((task) => ({ task, plugin: hook.plugin.id }))
      )
    }))
  )
  error.details = ''

  if (
    hookConflicts.length > 0 ||
    definedHookTaskConflicts.length > 0 ||
    taskConflicts.length > 0 ||
    optionConflicts.length > 0
  ) {
    shouldThrow = true

    if (hookConflicts.length) {
      error.details += formatHookConflicts(hookConflicts)
    }

    if (definedHookTaskConflicts.length) {
      error.details += formatHookTaskConflicts(definedHookTaskConflicts)
    }

    if (taskConflicts.length) {
      error.details += formatTaskConflicts(taskConflicts)
    }

    if (optionConflicts.length) {
      error.details += formatOptionConflicts(optionConflicts)
    }
  }

  const configuredHookTasks = withoutConflicts(Object.values(config.hookTasks))
  const definedHookIds = new Set(Object.keys(config.hooks))
  const undefinedHookTasks = configuredHookTasks.filter((hookTask) => {
    // we only care about undefined hooks that were configured by the app, not default config from plugins
    const fromApp = hookTask.plugin.root === process.cwd()
    const hookDefined = definedHookIds.has(hookTask.id)
    return fromApp && !hookDefined
  })

  if (undefinedHookTasks.length > 0) {
    shouldThrow = true
    error.details += formatUndefinedHookTasks(undefinedHookTasks, Array.from(definedHookIds))
  }

  const unusedOptions = Object.entries(config.options)
    .filter(
      ([, option]) =>
        option && !isConflict(option) && !option.forPlugin && option.plugin.root === process.cwd()
    )
    .map(([id]) => id)
  if (unusedOptions.length > 0) {
    shouldThrow = true
    error.details += formatUnusedOptions(unusedOptions, Object.keys(config.plugins))
  }

  const missingTasks = configuredHookTasks
    .map((hook) => ({
      hook,
      tasks: hook.tasks.filter((id) => !config.tasks[id])
    }))
    .filter(({ tasks }) => tasks.length > 0)

  if (missingTasks.length > 0) {
    shouldThrow = true
    error.details += formatMissingTasks(missingTasks, Object.keys(config.tasks))
  }

  if (shouldThrow) {
    throw error
  }
}

export function validatePlugins(config: RawConfig): Validated<ValidPluginsConfig> {
  const validatedPlugins = reduceValidated(
    Object.entries(config.plugins).map(([id, plugin]) => mapValidated(plugin, (p) => [id, p] as const))
  )
  return mapValidated(validatedPlugins, (plugins) => ({ ...config, plugins: Object.fromEntries(plugins) }))
}

export async function checkInstall(config: ValidConfig): Promise<void> {
  const definedHooks = withoutConflicts(Object.values(config.hooks))
  const uninstalledHooks = await asyncFilter(definedHooks, async (hook) => {
    return !(await hook.check())
  })

  if (uninstalledHooks.length > 0) {
    const error = new ToolKitError('There are problems with your Tool Kit installation.')
    error.details = formatUninstalledHooks(uninstalledHooks)
    throw error
  }
}

export function loadConfig(logger: Logger, options?: { validate?: true }): Promise<ValidConfig>
export function loadConfig(logger: Logger, options?: { validate?: false }): Promise<RawConfig>

export async function loadConfig(logger: Logger, { validate = true } = {}): Promise<ValidConfig | RawConfig> {
  const config = createConfig()

  // start loading config and child plugins, starting from the consumer app directory
  const rootPlugin = await loadPlugin('app root', config, logger)
  if (!rootPlugin.valid) {
    const error = new ToolKitError('root plugin was not valid!')
    error.details = rootPlugin.reasons.join('\n\n')
    throw error
  }
  const validRootPlugin = rootPlugin.value

  const validatedPluginConfig = validatePlugins(config)

  if (!validatedPluginConfig.valid) {
    const error = new ToolKitError('config was not valid!')
    error.details = validatedPluginConfig.reasons.join('\n\n')
    throw error
  }
  const validPluginConfig = validatedPluginConfig.value

  // collate root plugin and descendent hooks, options etc into config
  resolvePlugin(validRootPlugin, validPluginConfig, logger)

  if (validate) {
    validateConfig(validPluginConfig)
    return validPluginConfig
  }

  return config
}
