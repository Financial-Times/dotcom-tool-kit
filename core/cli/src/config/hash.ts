import { ValidConfig } from '@dotcom-tool-kit/config'
import { readState, writeState } from '@dotcom-tool-kit/state'
import { createHash } from 'node:crypto'
import { readFile } from 'node:fs/promises'
import { Logger } from 'winston'

export async function fileHash(path: string): Promise<string> {
  const hashFunc = createHash('sha512')
  try {
    hashFunc.update(await readFile(path))
    return hashFunc.digest('base64')
  } catch (error) {
    if ((error as NodeJS.ErrnoException)?.code === 'ENOENT') {
      return 'n/a'
    } else {
      throw error
    }
  }
}

export async function updateHashes(config: ValidConfig): Promise<void> {
  const hashes = Object.fromEntries(
    await Promise.all(
      ['.toolkitrc.yml', ...config.hookManagedFiles].map(async (path) => [path, await fileHash(path)])
    )
  )
  writeState('install', hashes)
}

export async function hasConfigChanged(logger: Logger, config: ValidConfig): Promise<boolean> {
  const hashes = readState('install')

  if (!hashes) {
    return true
  }

  for (const path of ['.toolkitrc.yml', ...config.hookManagedFiles]) {
    const newHash = await fileHash(path)
    const prevHash = hashes[path]

    if (newHash !== prevHash) {
      logger.debug(`hash for path ${path} has changed, running hook checks`)
      return true
    }
  }
  return false
}
