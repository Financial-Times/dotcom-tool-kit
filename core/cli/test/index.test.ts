import { ToolKitError } from '@dotcom-tool-kit/error'
import type { Valid } from '@dotcom-tool-kit/validated'
import type { Plugin } from '@dotcom-tool-kit/plugin'
import type { ValidPluginsConfig } from '@dotcom-tool-kit/config'
import { describe, expect, it, jest } from '@jest/globals'
import * as path from 'path'
import winston, { Logger } from 'winston'
import { createConfig, validateConfig } from '../src/config.js'
import { loadHookInstallations } from '../src/install.js'
import { loadPlugin, resolvePlugin, resolvePluginOptions } from '../src/plugin.js'
import { validatePlugins } from '../src/config/validate-plugins.js'

const logger = winston as unknown as Logger

// Loading all the plugins can (unfortunately) take longer than the default 2s timeout
jest.setTimeout(20000)

describe('cli', () => {
  // TODO:KB:202301121 we only return conflicts for hooks that are defined.
  // currently there are no hooks lol
  it.skip('should indicate when there are conflicts', async () => {
    const config = createConfig()

    const plugin = await loadPlugin('app root', config, logger, {
      id: 'conflicted test root',
      root: path.join(__dirname, 'files/conflicted')
    })
    expect(plugin.valid).toBe(true)

    const validatedPluginConfig = validatePlugins(config)
    expect(validatedPluginConfig.valid).toBe(true)
    const validPluginConfig = (validatedPluginConfig as Valid<ValidPluginsConfig>).value

    resolvePluginOptions((plugin as Valid<Plugin>).value, validPluginConfig)
    resolvePlugin((plugin as Valid<Plugin>).value, validPluginConfig, logger)

    expect(() => validateConfig(validPluginConfig, logger)).toThrow(ToolKitError)
    expect(validPluginConfig).toHaveProperty('commandTasks.build:ci.conflicting')
    expect(validPluginConfig).toHaveProperty('commandTasks.build:remote.conflicting')
    expect(validPluginConfig).toHaveProperty('commandTasks.build:local.conflicting')
  })

  it('should indicate when there are conflicts between plugins that are cousins in the tree', async () => {
    const config = createConfig()

    const plugin = await loadPlugin('app root', config, logger, {
      id: 'conflicted cousin test root',
      root: path.join(__dirname, 'files/cousins')
    })
    expect(plugin.valid).toBe(true)

    const validatedPluginConfig = validatePlugins(config)
    expect(validatedPluginConfig.valid).toBe(true)
    const validPluginConfig = (validatedPluginConfig as Valid<ValidPluginsConfig>).value

    resolvePluginOptions((plugin as Valid<Plugin>).value, validPluginConfig)
    resolvePlugin((plugin as Valid<Plugin>).value, validPluginConfig, logger)

    expect(() => validateConfig(validPluginConfig, logger)).toThrow(ToolKitError)
    expect(config).toHaveProperty('commandTasks.build:ci.conflicting')
    expect(config).toHaveProperty('commandTasks.build:remote.conflicting')
    expect(config).toHaveProperty('commandTasks.build:local.conflicting')
  })

  it('should not have conflicts between multiple of the same plugin', async () => {
    const config = createConfig()

    const plugin = await loadPlugin('app root', config, logger, {
      id: 'resolved test root',
      root: path.join(__dirname, 'files/duplicate')
    })
    expect(plugin.valid).toBe(true)

    const validatedPluginConfig = validatePlugins(config)
    expect(validatedPluginConfig.valid).toBe(true)
    const validPluginConfig = (validatedPluginConfig as Valid<ValidPluginsConfig>).value

    resolvePluginOptions((plugin as Valid<Plugin>).value, validPluginConfig)
    resolvePlugin((plugin as Valid<Plugin>).value, validPluginConfig, logger)

    try {
      const validConfig = validateConfig(validPluginConfig, logger)
      expect(validConfig).not.toHaveProperty('hooks.build:local.conflicting')
    } catch (e) {
      if (e instanceof ToolKitError) {
        e.message += '\n' + e.details
      }

      throw e
    }
  })

  it('should succeed when conflicts are resolved', async () => {
    const config = createConfig()

    const plugin = await loadPlugin('app root', config, logger, {
      id: 'resolved test root',
      root: path.join(__dirname, 'files/conflict-resolution')
    })
    expect(plugin.valid).toBe(true)

    const validatedPluginConfig = validatePlugins(config)
    expect(validatedPluginConfig.valid).toBe(true)
    const validPluginConfig = (validatedPluginConfig as Valid<ValidPluginsConfig>).value

    resolvePluginOptions((plugin as Valid<Plugin>).value, validPluginConfig)
    resolvePlugin((plugin as Valid<Plugin>).value, validPluginConfig, logger)

    try {
      const validConfig = validateConfig(validPluginConfig, logger)

      expect(validConfig).not.toHaveProperty('commandTasks.build:local.conflicting')
      expect(validConfig.commandTasks['build:local'].tasks.map((task) => task.task)).toEqual([
        'Webpack',
        'Babel'
      ])
    } catch (e) {
      if (e instanceof ToolKitError) {
        e.message += '\n' + e.details
      }

      throw e
    }
  })

  it('should successfully install when options for different hooks are defined', async () => {
    const config = createConfig()

    const plugin = await loadPlugin('app root', config, logger, {
      id: 'reolved test root',
      root: path.join(__dirname, 'files/multiple-hook-options')
    })
    expect(plugin.valid).toBe(true)

    const validatedPluginConfig = validatePlugins(config)
    expect(validatedPluginConfig.valid).toBe(true)
    const validPluginConfig = (validatedPluginConfig as Valid<ValidPluginsConfig>).value

    resolvePlugin((plugin as Valid<Plugin>).value, validPluginConfig, logger)

    const validConfig = validateConfig(validPluginConfig, logger)
    const hooks = await loadHookInstallations(logger, validConfig)
    expect(hooks.valid).toBe(true)
  })
})
