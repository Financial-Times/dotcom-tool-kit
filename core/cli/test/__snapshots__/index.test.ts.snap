// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should load plugins correctly 1`] = `
Object {
  "hookTasks": Object {
    "cleanup:remote": Object {
      "id": "cleanup:remote",
      "plugin": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
      "tasks": Array [
        "NpmPrune",
      ],
    },
    "deploy:production": Object {
      "id": "deploy:production",
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "tasks": Array [
        "HerokuProduction",
      ],
    },
    "deploy:review": Object {
      "id": "deploy:review",
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "tasks": Array [
        "HerokuReview",
      ],
    },
    "deploy:staging": Object {
      "id": "deploy:staging",
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "tasks": Array [
        "HerokuStaging",
      ],
    },
    "teardown:staging": Object {
      "id": "teardown:staging",
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "tasks": Array [
        "HerokuTeardown",
      ],
    },
    "test:ci": Object {
      "id": "test:ci",
      "plugin": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "tasks": Array [
        "Mocha",
        "Eslint",
      ],
    },
    "test:local": Object {
      "id": "test:local",
      "plugin": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "tasks": Array [
        "Mocha",
        "Eslint",
      ],
    },
    "test:review": Object {
      "id": "test:review",
      "plugin": Object {
        "id": "../../../../../plugins/n-test",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/n-test/lib/index.js",
      },
      "tasks": Array [
        "NTest",
      ],
    },
    "test:staged": Object {
      "id": "test:staged",
      "plugin": Object {
        "id": "../../../../../plugins/eslint",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/eslint/lib/index.js",
      },
      "tasks": Array [
        "Eslint",
      ],
    },
    "test:staging": Object {
      "id": "test:staging",
      "plugin": Object {
        "id": "../../../../../plugins/n-test",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/n-test/lib/index.js",
      },
      "tasks": Array [
        "NTest",
      ],
    },
  },
  "hooks": Object {
    "build:ci": BuildCI {
      "addToNightly": true,
      "circleConfigPath": ".circleci/config.yml",
      "id": "build:ci",
      "job": "tool-kit/build",
      "jobOptions": Object {
        "requires": Array [
          "tool-kit/setup",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/circleci",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/circleci/lib/index.js",
      },
    },
    "build:local": BuildLocal {
      "field": "scripts",
      "hook": "build:local",
      "id": "build:local",
      "key": "build",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/npm",
        "parent": Object {
          "id": "@dotcom-tool-kit/heroku",
          "parent": Object {
            "id": "../../../../../plugins/circleci-heroku",
            "parent": Object {
              "id": "successful test root",
              "root": "core/cli/test/files/successful",
            },
            "root": "plugins/circleci-heroku/lib/index.js",
          },
          "root": "plugins/heroku/lib/index.js",
        },
        "root": "plugins/npm/lib/index.js",
      },
    },
    "build:remote": BuildRemote {
      "field": "scripts",
      "hook": "build:remote",
      "id": "build:remote",
      "key": "heroku-postbuild",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
    },
    "cleanup:remote": CleanupRemote {
      "field": "scripts",
      "hook": "cleanup:remote",
      "id": "cleanup:remote",
      "key": "heroku-postbuild",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
    },
    "deploy:production": DeployProduction {
      "circleConfigPath": ".circleci/config.yml",
      "id": "deploy:production",
      "job": "tool-kit/heroku-promote",
      "jobOptions": Object {
        "filters": Object {
          "branches": Object {
            "only": "main",
          },
        },
        "requires": Array [
          "tool-kit/e2e-test-staging",
          "tool-kit/test",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
    "deploy:review": DeployReview {
      "circleConfigPath": ".circleci/config.yml",
      "id": "deploy:review",
      "job": "tool-kit/heroku-provision",
      "jobOptions": Object {
        "filters": Object {
          "branches": Object {
            "ignore": "/(^renovate-.*|^nori/.*|^main)/",
          },
        },
        "requires": Array [
          "tool-kit/setup",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
    "deploy:staging": DeployStaging {
      "circleConfigPath": ".circleci/config.yml",
      "id": "deploy:staging",
      "job": "tool-kit/heroku-staging",
      "jobOptions": Object {
        "filters": Object {
          "branches": Object {
            "only": "main",
          },
        },
        "requires": Array [
          "tool-kit/setup",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
    "release:remote": ReleaseRemote {
      "field": "scripts",
      "hook": "release:remote",
      "id": "release:remote",
      "key": "heroku-postbuild",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
    },
    "run:local": RunLocal {
      "field": "scripts",
      "hook": "run:local",
      "id": "run:local",
      "key": "start",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/npm",
        "parent": Object {
          "id": "@dotcom-tool-kit/heroku",
          "parent": Object {
            "id": "../../../../../plugins/circleci-heroku",
            "parent": Object {
              "id": "successful test root",
              "root": "core/cli/test/files/successful",
            },
            "root": "plugins/circleci-heroku/lib/index.js",
          },
          "root": "plugins/heroku/lib/index.js",
        },
        "root": "plugins/npm/lib/index.js",
      },
    },
    "teardown:staging": TestStaging {
      "circleConfigPath": ".circleci/config.yml",
      "id": "teardown:staging",
      "job": "tool-kit/e2e-test-staging",
      "jobOptions": Object {
        "requires": Array [
          "tool-kit/heroku-staging",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
    "test:ci": TestCI {
      "addToNightly": true,
      "circleConfigPath": ".circleci/config.yml",
      "id": "test:ci",
      "job": "tool-kit/test",
      "jobOptions": Object {
        "requires": Array [
          "tool-kit/build",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/circleci",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/circleci/lib/index.js",
      },
    },
    "test:local": TestLocal {
      "field": "scripts",
      "hook": "test:local",
      "id": "test:local",
      "key": "test",
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "@dotcom-tool-kit/npm",
        "parent": Object {
          "id": "@dotcom-tool-kit/heroku",
          "parent": Object {
            "id": "../../../../../plugins/circleci-heroku",
            "parent": Object {
              "id": "successful test root",
              "root": "core/cli/test/files/successful",
            },
            "root": "plugins/circleci-heroku/lib/index.js",
          },
          "root": "plugins/heroku/lib/index.js",
        },
        "root": "plugins/npm/lib/index.js",
      },
    },
    "test:review": TestReview {
      "circleConfigPath": ".circleci/config.yml",
      "id": "test:review",
      "job": "tool-kit/e2e-test-review",
      "jobOptions": Object {
        "requires": Array [
          "tool-kit/heroku-provision",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
    "test:staging": TestStaging {
      "circleConfigPath": ".circleci/config.yml",
      "id": "test:staging",
      "job": "tool-kit/e2e-test-staging",
      "jobOptions": Object {
        "requires": Array [
          "tool-kit/heroku-staging",
        ],
      },
      "logger": DerivedLogger {},
      "plugin": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
    },
  },
  "options": Object {
    "../../../../../plugins/eslint": Object {
      "forPlugin": Object {
        "id": "../../../../../plugins/eslint",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/eslint/lib/index.js",
      },
      "options": Object {
        "files": Array [
          "webpack.config.js",
        ],
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
    },
    "../../../../../plugins/mocha": Object {
      "forPlugin": Object {
        "id": "../../../../../plugins/mocha",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/mocha/lib/index.js",
      },
      "options": Object {
        "testDir": "./mocha_tests",
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
    },
    "../../../../../plugins/n-test": Object {
      "forPlugin": Object {
        "id": "../../../../../plugins/n-test",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/n-test/lib/index.js",
      },
      "options": Object {
        "host": "https://example.com",
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
    },
    "@dotcom-tool-kit/eslint": Object {
      "forPlugin": undefined,
      "options": Object {
        "files": "**/*.js",
      },
      "plugin": Object {
        "id": "../../../../../plugins/eslint",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/eslint/lib/index.js",
      },
    },
    "@dotcom-tool-kit/mocha": Object {
      "forPlugin": undefined,
      "options": Object {
        "files": "test/**/*.js",
      },
      "plugin": Object {
        "id": "../../../../../plugins/mocha",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/mocha/lib/index.js",
      },
    },
  },
  "plugins": Object {
    "../../../../../plugins/circleci-heroku": Object {
      "id": "../../../../../plugins/circleci-heroku",
      "parent": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "root": "plugins/circleci-heroku/lib/index.js",
    },
    "../../../../../plugins/eslint": Object {
      "id": "../../../../../plugins/eslint",
      "parent": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "root": "plugins/eslint/lib/index.js",
    },
    "../../../../../plugins/mocha": Object {
      "id": "../../../../../plugins/mocha",
      "parent": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "root": "plugins/mocha/lib/index.js",
    },
    "../../../../../plugins/n-test": Object {
      "id": "../../../../../plugins/n-test",
      "parent": Object {
        "id": "successful test root",
        "root": "core/cli/test/files/successful",
      },
      "root": "plugins/n-test/lib/index.js",
    },
    "@dotcom-tool-kit/circleci": Object {
      "id": "@dotcom-tool-kit/circleci",
      "parent": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "root": "plugins/circleci/lib/index.js",
    },
    "@dotcom-tool-kit/heroku": Object {
      "id": "@dotcom-tool-kit/heroku",
      "parent": Object {
        "id": "../../../../../plugins/circleci-heroku",
        "parent": Object {
          "id": "successful test root",
          "root": "core/cli/test/files/successful",
        },
        "root": "plugins/circleci-heroku/lib/index.js",
      },
      "root": "plugins/heroku/lib/index.js",
    },
    "@dotcom-tool-kit/npm": Object {
      "id": "@dotcom-tool-kit/npm",
      "parent": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
      "root": "plugins/npm/lib/index.js",
    },
    "@dotcom-tool-kit/vault": Object {
      "id": "@dotcom-tool-kit/vault",
      "parent": Object {
        "id": "@dotcom-tool-kit/heroku",
        "parent": Object {
          "id": "../../../../../plugins/circleci-heroku",
          "parent": Object {
            "id": "successful test root",
            "root": "core/cli/test/files/successful",
          },
          "root": "plugins/circleci-heroku/lib/index.js",
        },
        "root": "plugins/heroku/lib/index.js",
      },
      "root": "lib/vault/lib/index.js",
    },
  },
  "root": "core/cli",
  "tasks": Object {
    "Eslint": [Function],
    "HerokuProduction": [Function],
    "HerokuReview": [Function],
    "HerokuStaging": [Function],
    "HerokuTeardown": [Function],
    "Mocha": [Function],
    "NTest": [Function],
    "NpmPrune": [Function],
    "NpmPublish": [Function],
  },
}
`;
